[Feat] Implement room mention actionable whisper

fedirjh opened the pull request:
### Details

cc @rlinoz @rushatgabhane

### Fixed Issues

$ https://github.com/Expensify/App/issues/39508
https://github.com/Expensify/App/issues/39546

### Tests

1. Open App and navigate to any chat room
2. Create a new thread
3. Tag a non-existing user in the thread
4. Verify a mention whisper is displayed
5. Copy the thread id and run this code in the chrome console 
```jsx
async function convertMentionToReport(reportId) {
  const actions = await Onyx.get('reportActions_'+reportId);
  const action = Object.values(actions).filter(({actionName}) => actionName === 'ACTIONABLEMENTIONWHISPER')
  if (action.length) {
    Onyx.merge('reportActions_' + reportId, {
      [action[0].reportActionID]: {
        actionName: 'ACTIONABLEREPORTMENTIONWHISPER'
      }});
  }
}
await convertMentionToReport(id) // thread id from step 5
```
7. Verify the whisper buttons are updated
8. Press `create it` and verify there is a command call to `ResolveActionableReportMentionWhisper`
9. Verify that the whisper disappear

- [x] Verify that no errors appear in the JS console

### Offline tests

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>
<img width="458" alt="Screenshot 2024-05-01 at 1 19 14 PM" src="https://github.com/Expensify/App/assets/36869046/3cffefea-3ee3-48c2-93b7-8b2ebfe19ac5">

</details>

<details>
<summary>Android: mWeb Chrome</summary>
<img width="458" alt="Screenshot 2024-05-01 at 1 16 47 PM" src="https://github.com/Expensify/App/assets/36869046/512ad62b-9c37-474a-b09b-a16a61534abe">

</details>

<details>
<summary>iOS: Native</summary>
<img width="718" alt="Screenshot 2024-05-01 at 1 14 27 PM" src="https://github.com/Expensify/App/assets/36869046/9659a575-8c2f-455c-bca7-26cb5d3b2f4a">

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<img width="718" alt="Screenshot 2024-05-01 at 1 06 28 PM" src="https://github.com/Expensify/App/assets/36869046/a09306a1-a0cc-47f1-b5cc-a7f92db5a288">

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<img width="1182" alt="Screenshot 2024-05-01 at 10 22 11 AM" src="https://github.com/Expensify/App/assets/36869046/844b1bcf-f0c4-4f64-beb2-2118c63b9c84">

https://github.com/Expensify/App/assets/36869046/18e4fd98-d70a-4616-bb27-8e506890440a

</details>

<details>
<summary>MacOS: Desktop</summary>

<img width="1312" alt="Screenshot 2024-05-01 at 1 22 12 PM" src="https://github.com/Expensify/App/assets/36869046/9ab58ff0-4143-42d2-acaf-d153f2b3f475">

https://github.com/Expensify/App/assets/36869046/8368552f-e6b9-40c1-bc7d-4cc0d5a07863

</details>

fedirjh committed:
[Chore]: Add const and types for ACTIONABLE_ROOM_MENTION_WHISPER

fedirjh committed:
[Chore]: Add ResolveActionableRoomMentionWhisper Command

fedirjh committed:
[Chore]: Add isActionableRoomMentionWhisper function and options for handling actionable room mention whispers.

fedirjh committed:
[Refactor] Simplified the 'resolveActionableRoomMentionWhisper' function in 'Report.ts' by removing redundant message data.

fedirjh committed:
[Refactor] Improve handling of actionable whispers in report actions

This commit refactors the functions and conditions related to actionable whispers in report actions. The changes aim to improve the code's maintainability by reducing complexity and improving readability, which will make future modifications to this part of the codebase easier.

fedirjh committed:
[Chore] Add ResolveActionableRoomMentionWhisperParams to WRITE_COMMANDS.

rlinoz commented:
Sorry, I updated the GH but forgot about the doc, can we rename to ACTIONABLE_**REPORT**_MENTION_WHISPER instead of ROOM?

fedirjh committed:
[Refactor]: Rename 'RoomMention' components and methods to 'ReportMention'.

This change occurred in various files across the application, including action handlers, type definitions, API parameters, and translations.

rushatgabhane commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

rlinoz commented:
Just a quick updated, I am finishing the automated tests in the backend then I will provide a .env so we can test this properly.

fedirjh committed:
[Chore]: Remove redundant code

rlinoz commented on src/pages/home/report/ReportActionItem.tsx:
```diff
@@ -861,13 +870,8 @@ function ReportActionItem({
-
-    // if action is actionable track expense whisper and resolved by user, then we don't want to render anything
-    if (ReportActionsUtils.isActionableTrackExpense(action) && (action.originalMessage as OriginalMessageActionableTrackedExpenseWhisper['originalMessage']).resolution) {
+    // if action is actionable mention whisper or track expense whisper and resolved by user, then we don't want to render anything
```
```suggestion
    // If action is actionable whisper and resolved by user, then we don't want to render anything
```

rlinoz reviewed (changes requested):
It is working pretty well, just some minor things.

rlinoz commented on src/languages/es.ts:
```diff
@@ -3194,6 +3194,10 @@ export default {
         nothing: 'No hacer nada',
     },
+    actionableReportMentionWhisperOptions: {
+        create: 'Créalo',
```
Also, did we ask for a translation for this case?

fedirjh committed:
[Refactor]: Update action resolutions in ReportActionItem and CONST files

Replaced 'YES' and 'NO' resolutions for actionable report mentions with 'CREATE' and 'NOTHING' respectively. This improves the clarity of the mentioned actions across the application.

fedirjh committed:
Merge branch 'main' into room-mention-actionable-whisper

rlinoz commented on src/languages/en.ts:
```diff
@@ -2756,6 +2756,10 @@ export default {
         nothing: 'Do nothing',
     },
+    actionableReportMentionWhisperOptions: {
+        create: 'Create it',
+        nothing: 'Do nothing',
```
Right now the copy of the whisper is `Heads up, #qp doesn't exist yet. Do you want to create it?`. So I think it makes sense to change this to `Yes` and `No` for now.

We probably can use the ones we already have here https://github.com/Expensify/App/blob/main/src/languages/es.ts#L111-L112

rlinoz reviewed (changes requested):
Just one more update, other than that I think we are good!

fedirjh committed:
[Chore]: Refactor actionable report mention whisper options

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9011216944) to production by https://github.com/marcaaron in version: 1.4.71-6 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
